src/Datadog.Trace.ClrProfiler.Native/stats.h

```diff
@@ -0,0 +1,118 @@
+#ifndef DD_CLR_PROFILER_STATS_H_
+#define DD_CLR_PROFILER_STATS_H_
+
+#include "util.h"
+#include <chrono>
+
+namespace trace {
+
+class SWStat {
+  std::chrono::nanoseconds *_value;
+  std::chrono::steady_clock::time_point _startTime;
+
+ public:
+  SWStat(std::chrono::nanoseconds *value) {
+      _value = value;
+      _startTime = std::chrono::steady_clock::now();
+  }
+  ~SWStat() {
+    *_value += std::chrono::steady_clock::now() - _startTime;
+  }
+};
+
+class Stats : public Singleton<Stats> {
+  friend class Singleton<Stats>;
+
+ private:
+  std::chrono::nanoseconds callTargetRequestRejit;
+  std::chrono::nanoseconds callTargetRewriter;
+  std::chrono::nanoseconds jitInlining;
+  std::chrono::nanoseconds jitCompilationStarted;
+  std::chrono::nanoseconds moduleUnloadStarted;
+  std::chrono::nanoseconds moduleLoadFinished;
+  std::chrono::nanoseconds assemblyLoadFinished;
+  std::chrono::nanoseconds initialize;
+
+  //
+  unsigned int callTargetRequestRejitCount;
+  unsigned int callTargetRewriterCount;
+  unsigned int jitInliningCount;
+  unsigned int jitCompilationStartedCount;
+  unsigned int moduleUnloadStartedCount;
+  unsigned int moduleLoadFinishedCount;
+  unsigned int assemblyLoadFinishedCount;
+
+ public:
+  Stats() {
+    callTargetRequestRejit = std::chrono::nanoseconds(0);
+    jitInlining = std::chrono::nanoseconds(0);
+    jitCompilationStarted = std::chrono::nanoseconds(0);
+    moduleUnloadStarted = std::chrono::nanoseconds(0);
+    moduleLoadFinished = std::chrono::nanoseconds(0);
+    assemblyLoadFinished = std::chrono::nanoseconds(0);
+    initialize = std::chrono::nanoseconds(0);
+
+    callTargetRequestRejitCount = 0;
+    jitInliningCount = 0;
+    jitCompilationStartedCount = 0;
+    moduleUnloadStartedCount = 0;
+    moduleLoadFinishedCount = 0;
+    assemblyLoadFinishedCount = 0;
+  }
+  SWStat CallTargetRequestRejitMeasure() {
+    callTargetRequestRejitCount++;
+    return SWStat(&callTargetRequestRejit);
```

Comment:
Something I realised I've never checked - is the native profiler single threaded? I assume it must be (or we enforce that through queueing etc), but if not, these counters aren't thread safe right?

Reply:
Is not single threaded, callbacks can run from multiple threads, but we enforce syncronization with a lock guards in the callbacks, ex:
https://github.com/DataDog/dd-trace-dotnet/blob/32973349544145eb43dddefb6976ea5d0d4b206b/src/Datadog.Trace.ClrProfiler.Native/cor_profiler.cpp#L293-L295

But you're right the guard is after creating this SWStat instance, I will add synchronisation here too.